spec-terminology.md

```diff
@@ -1,30 +1,75 @@
 # Terminology about specifications
 
-Language and platform specifications have several different terms used to
-describe how well-defined a language feature is, i.e., how well constrained the
-runtime behavior is.
+Language and platform specifications have several different terms used
+to describe how well-defined a language feature is, i.e., how well
+constrained the runtime behavior is. In cases where our terminology
+overlaps with that from other communities, we try to remain generally
+compatible.
 
-The ISO C and C++ standards distinguish several degrees of specification by
-assigning precise definitions to the following terms:
+<a name="UB"></a>
 
-* Well-defined behavior
-* Implementation-defined behavior
-* Unspecified behavior
-* Undefined behavior
+## Undefined Behavior
 
-Of these, only "undefined behavior" is used consistently within the Rust
-project; there are _no_ guarantees about the runtime behavior of such code.
+As is typical within the Rust community we use the phrase **undefined
+behavior** to refer to illegal program actions that can result in
+arbitrary results. In short, "undefined behavior" is always a bug and
+never something you should do. See the [Rust
+reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)
+for more details.
 
-It is currently out of scope for this project to define the other terms in that
-list or their relationship to the corresponding terms in other languages.
-However, this project does distinguish a particular _category_ of undefined
-behavior:
+Our usage of the term is generally the same as the [standard
+usage](https://en.wikipedia.org/wiki/Undefined_behavior) from other
+languages.
 
-#### LLVM-undefined behavior (or LLVM-UB)
+<a name="LLVM-UB"></a>
 
-Cases of known LLVM-UB are a specific subset of undefined behavior in general.
-Rust code with LLVM-UB will cause `rustc` to generate LLVM IR exhibiting
-undefined behavior.
+### LLVM-undefined behavior (LLVM-UB)
 
-This is distinct from the general case of Rust undefined behavior, in which
-it is unknown whether `rustc` will generate well-behaved LLVM IR.
+As a special case of undefined behavior, we use the phrase **LLVM
```

Comment:
I think that the phrase "special case" implies that it is. However, I'm not sure how important the distinction really is at the end of the day.

That said, I debated about whether to include that phrase. I think it's probably worth distinguishing LLVM-UB as an orthogonal concept -- LLVM-UB is whatever LLVM defines as UB. Any case where Rust generates LLVM that exhibits LLVM-UB is a bug unless that behavior is also UB in Rust -- but those bugs can and do happen so it's good to be able to talk about it.

Reply:
> I think that the phrase "special case" implies that it is. 

I thought that the goal of the term "LLVM-UB" was to use it for those cases in which Rust has UB, but we guarantee that the generated LLVM-IR does not. For example, unwinding through "C unwind" is Rust-UB because we don't want to guarantee what that does, but it is not LLVM-UB because we want people to use it. If "LLVM-UB" implies "Rust-UB" then by saying that "C unwind" is not "LLVM-UB" we are saying that it is not "Rust-UB". Does that make sense ?

Reply:
I think this concern is basically resolved now -- I've clarified that LLVM-UB just refers to what is UB in LLVM, which is (hopefully) a subset of Rust UB (but not necessarily).

Moreover, as per this document, unwinding through "C unwind" is not Rust UB, but it *is* unspecified (or, more precisely, "TBD").

Reply:
Resolving.